packages/fuels-signers/src/provider.rs

```diff
@@ -399,6 +455,74 @@ impl Provider {
         self.client.produce_block(amount).await
     }
 
+    /// Get estimated cost of a transaction
+    pub async fn get_transaction_cost(
+        &self,
+        tx: &Transaction,
+        tolerance: Option<f64>,
+    ) -> Result<TransactionCost, Error> {
+        let tolerance = tolerance.unwrap_or(DEFAULT_GAS_ESTIMATION_TOLERANCE);
+
+        let NodeInfo {
+            min_gas_price,
+            min_byte_price,
+            ..
+        } = &self.node_info().await?;
+
+        let gas_price = std::cmp::max(tx.gas_price(), min_gas_price.0);
+        let byte_price = std::cmp::max(tx.byte_price(), min_byte_price.0);
+
+        // Simulate the contract call with MAX_GAS_PER_TX to get the complete gase_used
+        let mut tmp_tx = tx.clone();
+        tmp_tx.set_gas_limit(MAX_GAS_PER_TX);
+        tmp_tx.set_gas_price(0);
+        tmp_tx.set_byte_price(0);
+
+        let gas_used = self.get_gas_used_with_tolerance(&tmp_tx, tolerance).await?;
+        let byte_size = self.get_chargable_byte_size(tmp_tx);
+        // GAS_PRICE_FACTOR is a chaing_config of the  node. Because of the different decimal precision in
+        // FuelVM and EVM we need to scale the price down
+        let gas_fee = ((gas_used as f64 / GAS_PRICE_FACTOR as f64) * gas_price as f64) as u64;
+        let byte_fee = ((byte_size as f64 / GAS_PRICE_FACTOR as f64) * byte_price as f64) as u64;
+
+        Ok(TransactionCost {
+            min_gas_price: min_gas_price.0,
+            min_byte_price: min_byte_price.0,
+            byte_price,
+            gas_price,
+            gas_used,
+            byte_size,
+            fee: gas_fee + byte_fee,
+        })
+    }
```

Comment:
Whenever possible try to separate 'calculations' (pure functions) from 'actions' (side-effect inducing functions). Doing so will aid in testing. The thing about actions is that whichever function touches it - it itself becomes an action (impure). Something like the following is just a 5 min go at it. You can now easily unit-test the calculating logic without having to mock the heck out of the provider. 
```suggestion
/// Get estimated cost of a transaction
    pub async fn get_transaction_cost(
        &self,
        tx: &Transaction,
        tolerance: Option<f64>,
    ) -> Result<TransactionCost, Error> {
        let NodeInfo {
            min_gas_price,
            min_byte_price,
            ..
        } = self.node_info().await?;

        let tolerance = tolerance.unwrap_or(DEFAULT_GAS_ESTIMATION_TOLERANCE);
        let tmp_tx = Self::generate_dry_run_tx(tx);
        let gas_used = self.get_gas_used_with_tolerance(&tmp_tx, tolerance).await?;

        Self::calculate_tx_cost(tx, tmp_tx, min_gas_price, min_byte_price, gas_used)
    }

    fn generate_dry_run_tx(tx: &Transaction) -> Transaction {
        let mut dry_run_tx = tx.clone();
        dry_run_tx.set_gas_limit(MAX_GAS_PER_TX);
        dry_run_tx.set_gas_price(0);
        dry_run_tx.set_byte_price(0);
        dry_run_tx
    }

    fn calculate_tx_cost(
        tx: &Transaction,
        dry_run_tx: Transaction,
        min_gas_price: U64,
        min_byte_price: U64,
        gas_used: u64,
    ) -> Result<TransactionCost, Error> {
        let gas_price = std::cmp::max(tx.gas_price(), min_gas_price.0);
        let byte_price = std::cmp::max(tx.byte_price(), min_byte_price.0);

        // Simulate the contract call with MAX_GAS_PER_TX to get the complete gas_used

        let byte_size = Self::get_chargable_byte_size(dry_run_tx);
        // GAS_PRICE_FACTOR is a chain_config of the  node. Because of the different decimal precision in
        // FuelVM and EVM we need to scale the price down
        let gas_fee = ((gas_used as f64 / GAS_PRICE_FACTOR as f64) * gas_price as f64) as u64;
        let byte_fee = ((byte_size as f64 / GAS_PRICE_FACTOR as f64) * byte_price as f64) as u64;

        Ok(TransactionCost {
            min_gas_price: min_gas_price.0,
            min_byte_price: min_byte_price.0,
            byte_price,
            gas_price,
            gas_used,
            byte_size,
            fee: gas_fee + byte_fee,
        })
    }
```